db/commitlog/commitlog.cc

```diff
@@ -2303,6 +2372,9 @@ void db::commitlog::segment_manager::on_timer() {
                 flush_segments(cur + extra - max);
             }
         }
+
+        check_no_data_older_than_allowed();
```

Comment:
why is this done outside the gate?
It spawn background fibers (`s->close()`)
and it should wait for them so they won't race with shutdown.

Reply:
It is inside the "with_gate". I think the diff is tricking you.

Reply:
ok, but it's bogus since it doesn't return a future so it will still run stuff i the background after returning, losing the gate's protection.

`check_no_data_older_than_allowed` should collect it's background futures and return a future that awaits all of them, and it could be `co_await`ed here.

Reply:
No, it is not running anything in background (more than anything else that is). While I principally agree that dangling futures is bad, in this case it is how we do stuff (see other instances of "close()"). I.e. this is one path we simply send off in background, and it will resolve and sync via the gate eventually.

Reply:
segment::close() has its own protections.

Reply:
so why the `fixme. discarded future.` comment?
if there's nothing to fix the this comment should explain why it's safe.
we should mention instead that it's ok to discard the future and mention where/how we wait for it.

Reply:
Because dangling futures are imho still bad. In fact, I have a whole refactor PR partly to rid ourselves of them. But this PR is not the place. It is fine, in the sense it will not break things. But it mans we have unachored work, which I dislike in a ocd fashion.

Reply:
okay